### PR TITLE
Feat expire key

### PR DESCRIPTION
**Pull request by shimonewman**
## What does this PR do?

(Provide a description of what this PR does.)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work.)

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

(Write your answer here.)

### CODE REVIEW

File: app/controllers/api/projects.php [hunk 1]
```diff
@@ -775,9 +776,10 @@
     ->param('projectId', null, new UID(), 'Project unique ID.')
     ->param('name', null, new Text(128), 'Key name. Max length: 128 chars.')
     ->param('scopes', null, new ArrayList(new WhiteList(array_keys(Config::getParam('scopes')), true), APP_LIMIT_ARRAY_PARAMS_SIZE), 'Key scopes list. Maximum of ' . APP_LIMIT_ARRAY_PARAMS_SIZE . ' scopes are allowed.')
+    ->param('expire', 0, new Integer() , 'Key expiration time')
```
Comment (TorstenDittmann) [inline app/controllers/api/projects.php:779]: I think we can make this optional 👍🏻

File: app/controllers/general.php [hunk 2]
```diff
@@ -279,6 +279,13 @@
             $role = Auth::USER_ROLE_APP;
             $scopes = \array_merge($roles[$role]['scopes'], $key->getAttribute('scopes', []));
 
+            $expire = $key->getAttribute('expire', 0);
+
+            if($expire !== 0 && $expire < \time()){
+                throw new AppwriteException('Project key expired', 401, AppwriteException:: PROJECT_KEY_EXPIRED);
+             }
+
+            Authorization::setRole('role:'.Auth::USER_ROLE_APP);
```
Comment (TorstenDittmann) [inline app/controllers/general.php:288]:
```suggestion
```

Comment (TorstenDittmann) [reply app/controllers/general.php:288]: Looks like a copy pasta 😉

File: app/config/collections.php [hunk 3]
```diff
@@ -800,7 +800,7 @@
                 'size' => Database::LENGTH_KEY,
                 'signed' => true,
                 'required' => false,
-                'default' => null,
+                'default' => 0,
```
Comment (TorstenDittmann) [inline app/config/collections.php:803]: This default should stay `null` 🙂

File: app/controllers/api/projects.php (hunk 1 above)
Comment (TorstenDittmann) [inline app/controllers/api/projects.php:779]:
```suggestion
    ->param('expire', 0, new Integer(), 'Key expiration time')
```

File: src/Appwrite/Utopia/Response/Model/Key.php [hunk 4]
```diff
@@ -27,6 +27,12 @@ public function __construct()
                 'default' => '',
                 'example' => 'My API Key',
             ])
+            ->addRule('expire', [
+                'type' => self::TYPE_INTEGER,
+                'description' => 'Key expiration .',
```
Comment (TorstenDittmann) [inline src/Appwrite/Utopia/Response/Model/Key.php:32]:
```suggestion
                'description' => 'Key expiration timestamp.',
```

File: app/controllers/api/projects.php (hunk 1 above)
Comment (shimonewman) [reply app/controllers/api/projects.php]: You mean to pass TRUE in optional?

File: app/controllers/general.php (hunk 2 above)
Comment (shimonewman) [reply app/controllers/general.php:288]: (:  cool

File: app/config/collections.php (hunk 3 above)
Comment (shimonewman) [reply app/config/collections.php:803]: Cool

File: app/controllers/api/projects.php (hunk 1 above)
Comment (TorstenDittmann) [reply app/controllers/api/projects.php]: Exactly ☺️

Comment (shimonewman) [reply app/controllers/api/projects.php]: cool

File: app/config/collections.php (hunk 3 above)
Comment (TorstenDittmann) [reply app/config/collections.php:803]: Oh, the default of `projectId` needs to be null, the default of `expire` should be `0` 👍🏻

File: app/controllers/api/projects.php [hunk 5]
```diff
@@ -775,9 +776,10 @@
     ->param('projectId', null, new UID(), 'Project unique ID.')
     ->param('name', null, new Text(128), 'Key name. Max length: 128 chars.')
     ->param('scopes', null, new ArrayList(new WhiteList(array_keys(Config::getParam('scopes')), true), APP_LIMIT_ARRAY_PARAMS_SIZE), 'Key scopes list. Maximum of ' . APP_LIMIT_ARRAY_PARAMS_SIZE . ' scopes are allowed.')
+    ->param('expire', 0, new Integer() , 'Key expiration timestamp', true)
```
Comment (eldadfux) [inline app/controllers/api/projects.php:779]:
```suggestion
    ->param('expire', 0, new Integer() , 'Key expiration time in Unix timestamp.', true)
```

Very important to choose the right wording, the description is later used on the official docs.

File: src/Appwrite/Utopia/Response/Model/Key.php [hunk 6]
```diff
@@ -27,6 +27,12 @@ public function __construct()
                 'default' => '',
                 'example' => 'My API Key',
             ])
+            ->addRule('expire', [
+                'type' => self::TYPE_INTEGER,
+                'description' => 'Key expiration timestamp.',
```
Comment (eldadfux) [inline src/Appwrite/Utopia/Response/Model/Key.php:32]:
```suggestion
                'description' => 'Key expiration in Unix timestamp.',
```

Again, very important to be very accurate in the descriptions. These are later used in the docs.

File: app/controllers/api/projects.php [hunk 7]
```diff
@@ -775,9 +776,10 @@
     ->param('projectId', null, new UID(), 'Project unique ID.')
     ->param('name', null, new Text(128), 'Key name. Max length: 128 chars.')
     ->param('scopes', null, new ArrayList(new WhiteList(array_keys(Config::getParam('scopes')), true), APP_LIMIT_ARRAY_PARAMS_SIZE), 'Key scopes list. Maximum of ' . APP_LIMIT_ARRAY_PARAMS_SIZE . ' scopes are allowed.')
+    ->param('expire', 0, new Integer() , 'Key expiration time in Unix timestamp.', true)
```
Comment (eldadfux) [inline app/controllers/api/projects.php:779]:
```suggestion
    ->param('expire', 0, new Integer() , 'Key expiration time in Unix timestamp. Use 0 for unlimited expiry.', true)
```

Another small addition

**Pull request merged**